AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +604,842 @@ class AzureIoTHub {
+                mqttMsg.sendasync(callback);
+            } else {
+                local err = _state & S_CONNECTED ? E_OP_NOT_ALLOWED_NOW : E_NOT_CONNECTED;
+                _callOnComplete(onComplete, err);
             }
         }
 
-        // Note: auth receiver will open auth sender if needed, auth
-        function _amqpAuthSenderStatusHandler(event, data) {
-            _log("Auth Sender event: " + event);
-            switch(event) {
-                case "SENDER_OPEN":
-                    _getPutToken( function() { _log("put token request sent") });
-                    break;
-                case "SENDER_CLOSED":
-                    break;
-                case "SENDER_ERROR":
-                    if (_connecting) {
-                        if (data == null) data = event;
-                        _connectionCallback(data);
-                    } else {
-                        // TODO: replace this with error handling
-                        _log(data);
-                    }
-                    break;
-                default:
-                    _log(data);
-                    break;
-            }
-        }
+        // Enables or disables message receiving from Azure IoT Hub (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#receiving-cloud-to-device-messages).
+        //
+        // Parameters:
+        //     onReceive : Function         Callback called every time a new message is received. null disables the feature.
+        //                                  The callback signature:
+        //                                  onReceive(message), where
+        //                                      message :           Received message.
+        //                                          AzureIoTHub.Message
+        //     onComplete : Function        Callback called when the operation is completed or an error happens.
+        //          (optional)              The callback signature:
+        //                                  onComplete(error), where
+        //                                      error : Integer     0 if the operation is completed successfully, an error code otherwise.
+        //
+        // Returns:                         Nothing.
+        function enableMessageReceiving(onReceive, onComplete = null){
+            if ((_state & S_CONNECTED) && !(_state & S_DISCONNECTING)) {
+                if (_state & (S_ENABLING_MSG | S_DISABLING_MSG)) {
+                    _callOnComplete(onComplete, E_OP_NOT_ALLOWED_NOW);
+                } else if (_onMessageCb != null && onReceive == null) {
+                    // Should unsubscribe
+                    // TODO: Make sure we have successfully unsubscribed
+                    // TODO: Handle the case with disconnection when unsubscribing
+                    // TODO: Will be done once EI make callback for unsubscribe method
+                    _state = _state ^ S_DISABLING_MSG;
```

Comment:
1) you probably meant to use `|` instead of `^` here
2) why do you need it at all? it's a single threaded execution environment...